Content.Shared/Emag/Systems/EmagSystem.cs

```diff
@@ -0,0 +1,167 @@
+        }
+
+        private void OnExamine(EntityUid uid, EmagComponent component, ExaminedEvent args)
+        {
+            args.PushMarkup(Loc.GetString("emag-charges-remaining", ("charges", component.Charges)));
+            if (component.Charges == component.MaxCharges)
+            {
+                args.PushMarkup(Loc.GetString("emag-max-charges"));
+                return;
+            }
+            var timeRemaining = Math.Round((component.NextChargeTime - _timing.CurTime).TotalSeconds);
+            args.PushMarkup(Loc.GetString("emag-recharging", ("seconds", timeRemaining)));
+        }
+
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            foreach (var emag in EntityQuery<EmagComponent>())
+            {
+                if (!emag.AutoRecharge)
+                    continue;
+
+                if (emag.Charges == emag.MaxCharges)
+                    continue;
+
+                if (_timing.CurTime < emag.NextChargeTime)
+                    continue;
+
+                ChangeEmagCharge(emag.Owner, 1, true, emag);
+            }
+        }
+
+        private void OnAfterInteract(EntityUid uid, EmagComponent component, AfterInteractEvent args)
+        {
+            if (!args.CanReach || args.Target is not { } target)
+                return;
+
+            args.Handled = TryUseEmag(uid, args.User, target, component);
+        }
+
+        /// <summary>
+        /// Changes the charge on an emag.
+        /// </summary>
+        public bool ChangeEmagCharge(EntityUid uid, int change, bool resetTimer, EmagComponent? component = null)
+        {
+            if (!Resolve(uid, ref component))
+                return false;
+
+            if (component.Charges + change < 0 || component.Charges + change > component.MaxCharges)
+                return false;
+
+            if (resetTimer || component.Charges == component.MaxCharges)
+                component.NextChargeTime = _timing.CurTime + component.RechargeDuration;
+
+            component.Charges += change;
+            Dirty(component);
+            return true;
+        }
+
+        /// <summary>
+        /// Tries to use the emag on a target entity
+        /// </summary>
+        public bool TryUseEmag(EntityUid emag, EntityUid user, EntityUid target, EmagComponent? component = null)
+        {
+            if (!Resolve(emag, ref component, false))
+                return false;
+
+            if (_tagSystem.HasTag(target, component.EmagImmuneTag))
+                return false;
+
+            if (component.Charges <= 0)
+            {
+                _popupSystem.PopupEntity(Loc.GetString("emag-no-charges"), user, user);
+                return false;
+            }
+
+            var handled = DoEmagEffect(user, target);
+            if (!handled)
+                return false;
+
+            // only do popup on client
+            if (_net.IsClient && _timing.IsFirstTimePredicted)
```

Comment:
is IsFirstTimePredicted true on server?

Reply:
does it matter, considering it only runs on the client?

Reply:
Was wondering whether .IsClient was necessary.

Reply:
pretty sure it does.